lib/hanami/container.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+require "dry/system/container"
+
+module Hanami
+  # Hanami private IoC
+  #
+  # @since 2.0.0
+  class Container < Dry::System::Container
+    boot(:root) do
+      start do
+        register(:root, Hanami.root)
+      end
+    end
+
+    boot(:env) do |c|
+      init do
+        use :root
+
+        begin
+          require "dotenv"
+        rescue LoadError # rubocop:disable Lint/HandleExceptions
+        end
+
+        Dotenv.load(c[:root].join(".env")) if defined?(Dotenv)
```

Comment:
Having pluggable settings providers certainly _sounds_ like a nice thing to me. I guess it'll come down to whether we can maintain a consistent settings interface/experience while still allowing each settings provider's unique features to be leveraged.

What I'm definitely in favour of is having the settings explicitly listed (ideally along with their types), and then vended from a central object, which is what that dry-system settings component does. It's much cleaner than ENV and opens up possibilities for doing smarter things with settings, since we have a central place to manage them.

Happy for this to be a follow-up issue.

Reply:
@timriley Is that DRY mechanism that you linked a kind of replacement for `dotenv`? If so, does it support multiple envs and/or overrides via local files?

Reply:
@jodosha it supports environment-named files, but not the `.local` versions. We could extend it to include those, but perhaps a better option to have this dry-system plugin somehow abstract away the settings provider (like we discussed just above, here)

Reply:
@jodosha @timriley loading of env files is currently hard-coded [here](https://github.com/dry-rb/dry-system/blob/master/lib/dry/system/settings/file_loader.rb#L19-L24) but it would be very, very simple to add a configuration to `:settings` component that would specify file name hierarchy.

Reply:
@timriley & I had a long discussion regarding settings and env vars. We decided to leave the PR as it is, and to structure this feature in a future PR.